File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    private List<List<String>> res = new ArrayList<>();

    private boolean isPali(String s, int l, int r){
        while(l<r){
            if(s.charAt(l)!=s.charAt(r)){
                return false;
            }
            l++;
            r--;
        }

        return true;
    }

    private void backtrack(String s, int start, List<String> path) {
        if(start == s.length()){
            res.add(new ArrayList<>(path));
            return;
        }

        for(int end=start+1; end<=s.length(); end++){
            if(isPali(s, start, end-1)){
                path.add(s.substring(start,end));
                backtrack(s, end, path);
                path.remove(path.size()-1);
            }
        }
    }

    public List<List<String>> partition(String s) {
        backtrack(s, 0, new ArrayList<>());
        return res;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        n = len(s)
        res = []

        def isPalindrome(word):
            return word == word[::-1]

        def findPartition(start, curr):
            if start == n:
                res.append(list(curr))
                return 

            for end in range(start+1, n+1):
                if isPalindrome(s[start:end]):
                    findPartition(end, curr + [s[start:end]])
        
        findPartition(0, [])
        return res